docs/storage/storage.md

```diff
@@ -0,0 +1,186 @@
+yet fully populated.
+
+![strata depth 8 tree](StratumDepth8.png, "Stratum Depth 8")
+
+As it's hard to visualize the structure at scale with stratum depth 8 some examples of smaller
+depths might make things clearer, though these are not supported by the code the diagrams are
+much simpler.
+
+This diagram shows a tree with strata depth 2. It is a somewhat special case as all the levels are
+stored. Note that the root node is never stored and is always recalculated.
+
+![strata depth 2 tree diagram](StratumDepth2.png, "Stratum Depth 2")
+
+This diagram shows a tree with strata depth 3. Note that only the bottom level of each subtree is
+stored and how the binary path is used as a subtree prefix to identify subtrees.
+
+![strata depth 3 tree diagram](StratumDepth3.png, "Stratum Depth 3")
+
+### Consistency and Other Requirements
+
+Storage implementations must provide strongly consistent updates to the tree data. Some users may
+see an earlier view than others if updates have not been fully propagated yet but they must not see
+partial updates or inconsistent views.
+
+It is not a requirement that the underlying storage is relational. Our initial implementation uses
+an RDBMS and has this [database schema diagram](database-diagram.pdf, "Trillian DB Schema")
+
+## Map Storage
+
+Need stuff here.
+
+## Log Storage
+
+### The Log Tree
+
+The node tree built for a log is a representation of a Merkle Tree, which starts out empty and grows
+as leaves are added. A Merkle Tree of a specific size is a fixed and well-defined shape.
+                     
+Leaves are never removed and a completely filled in left subtree of the tree
+structure is never further mutated. 
+
+The log stores two hashes per leaf, a raw SHA256 hash of the leaf data used for deduplication and
+the Merkle Leaf Hash of the data, which becomes part of the tree.
+
+### Log NodeIDs / Tree Coordinates
+
+Log nodes are notionally addressed using a three dimensional coordinate tuple (level in tree, index
+in level, revision number).
+
+Level zero is always the leaf level and additional intermediate levels are added above this as the
+tree grows. Levels are only created when they are used.
+
+Index is the horizontal position of the node in the level, with zero being the leftmost node in
+each level.
+
+For example in a tree of size two the leaves are level 0 and index 0 and 1 and the root is
+level 1, index 0.
+
+The storage implementation must be able to provide access to nodes using this coordinate scheme but
+it is not required to store them this way. The current implementation collapses subtrees for
+increased write efficiency so nodes are not distinct database entities, this is hidden by the node
+API.
+
+### Log Startup
+
+When log storage is intialized and its tree is not empty the existing state is loaded into a
+`compact_merkle_tree`. This can be done efficiently and only requires a few node accesses. As a
```

Comment:
'a few node accesses' implies that you certainly don't need to read every leaf when calculating the new root hash using the compact merkle tree.

is that because it's possible to 'short circuit' that by adding the sub-tree roots at a particular instead, and have each 'contribute' the sum of all the descendant leaves?